### PR TITLE
Github Enterprise Support

### PR DESCRIPTION
**Pull request by gaborbernat**
support

This solves https://github.com/rtfd/readthedocs.org/issues/2194.

---
This change is [Reviewable](https://reviewable.io/reviews/rtfd/readthedocs.org/3853)

### CODE REVIEW

File: docs/custom_installs/github_enterprise.rst [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        from django.contrib.sites.models import Site
+        from django.conf import settings
+
+        # create a site for the authentication system
+        PRODUCTION_DOMAIN = getattr(settings, 'PRODUCTION_DOMAIN')
+        SITE_ID = getattr(settings, 'SITE_ID')
+        site, _ = Site.objects.update_or_create(id=SITE_ID, defaults=dict(name=PRODUCTION_DOMAIN, domain=PRODUCTION_DOMAIN))
+
+        # change the github social account provider to link to the enterprise Github
+        GITHUB = "{{github_url}}"
```
Comment (humitos) [inline docs/custom_installs/github_enterprise.rst]: This doesn't seem to be used at all.

**Comment by gaborbernat** [comment]
@humitos yes you need to change that, however in some places the ``github.com`` was hard coded into the code, I've replaced in these cases to instead read it from ``from allauth.socialaccount.providers.github.provider import GitHubProvider``  (also I've already mention you need to do the change in there at: https://github.com/gaborbernat/readthedocs.org/blame/38ee82afe90b2a0af455f93b7baa699db4b8cf50/docs/custom_installs/github_enterprise.rst#L32 - this is a config change) 
We could expose a main level variable in config if you guys think it would make it easier. 

To have the connect account feature work you need to setup the site setting correctly (e.g. point to your domain and not example.com as it's by default - which is the first change in the script). 

As for having both exists do we have a strong need for that, as would probably complicate the implementation.

**Comment by gaborbernat** [comment]
@humitos made the requested changes, ready to discuss supporting an enterprise instance with github.com in parallel.

File: readthedocs/builds/models.py [hunk 2]
```diff
@@ -5,6 +5,8 @@
     absolute_import, division, print_function, unicode_literals)
 
 import logging
+
+from allauth.socialaccount.providers.github.views import GitHubOAuth2Adapter
```
Comment (xrmx) [inline readthedocs/builds/models.py]: from should go after import, and third party import should go after stdlib import and before local imports

File: readthedocs/core/management/commands/import_github_language.py [hunk 3]
```diff
@@ -37,17 +40,12 @@ def handle(self, *args, **options):
             return
 
         for project in Project.objects.filter(
-            programming_language__in=['none', '', 'words']
+                programming_language__in=['none', '', 'words']
```
Comment (xrmx) [inline readthedocs/core/management/commands/import_github_language.py]: 4 spaces was fine

File: tox.ini [hunk 4]
```diff
@@ -15,6 +15,7 @@ setenv =
     DJANGO_SETTINGS_MODULE=readthedocs.settings.test
     LANG=C
     DJANGO_SETTINGS_SKIP_LOCAL=True
+passenv = https_proxy http_proxy no_proxy
```
Comment (xrmx) [inline tox.ini]: looks unrelated

File: readthedocs/core/management/commands/import_github_language.py [hunk 5]
```diff
@@ -11,13 +11,16 @@
 from __future__ import absolute_import
 from __future__ import print_function
 import os
+
+from allauth.socialaccount.providers.github.views import GitHubOAuth2Adapter
```
Comment (xrmx) [inline readthedocs/core/management/commands/import_github_language.py]: same comment as before

**Comment by gaborbernat** [comment]
Made the requested changes. By the way running ``isort`` would modify a lot of existing source files :thinking: are we sure it's configured correctly?

File: docs/custom_installs/github_enterprise.rst [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+Github integration is provided via the `django-allauth`_ package, and by default connected with `github.com`_.
+To swap out this connection with an enterprise edition of Github you need to:
+
+1. Change the social account provider settings to point the Enterprise Github (value of ``{{github_url}}`` in the
+   ``local_settings.py``):
+
+    .. code-block:: python
+
+        SOCIALACCOUNT_PROVIDERS = {'github': {'SCOPE': ['user:email', 'read:org',
```
[the remark this answers is not in the document]
Comment (humitos) [reply docs/custom_installs/github_enterprise.rst]: Also, if you modify `local_settings.py` like this, you override the setting and loose the `gitlab` feature, for example.

I've been using somethin like this:

```python
from .dev import CommunityDevSettings

INSTALLED_APPS = CommunityDevSettings().INSTALLED_APPS + [
    'debug_toolbar',
]
```

File: readthedocs/projects/constants.py [hunk 7]
```diff
@@ -291,10 +291,10 @@
 PROJECT_PK_REGEX = '(?:[-\w]+)'
 PROJECT_SLUG_REGEX = '(?:[-\w]+)'
 
-GITHUB_REGEXS = [
-    re.compile('github.com/(.+)/(.+)(?:\.git){1}$'),
-    re.compile('github.com/(.+)/(.+)'),
-    re.compile('github.com:(.+)/(.+)\.git$'),
+GITHUB_REGEX_BASES = [
+    '{}/(.+)/(.+)(?:\.git){{1}}$',
```
Comment (humitos) [inline readthedocs/projects/constants.py]: Not sure if this was on purpose, but it was `{1}` and now it's `{{1}}`. I suppose the original was correct.

**Comment by humitos** [comment]
> Made the requested changes. By the way running isort would modify a lot of existing source files thinking are we sure it's configured correctly?

Yes. It's correctly configured but it's something new and we didn't run to the entire project yet. We were running it only the files that we modify on each PR.

So, you could run it only on those files.

At a higher level, we have `pre-commit` configured (which will run only on your modified fields) but we are not forcing the community PR to run this command yet.

**Comment by humitos** [comment]
> As for having both exists do we have a strong need for that, as would probably complicate the implementation.

We don't really need this for the community site since we don't support GHE in .org but I was asking you to know if this could be applicable in the readthedocs.com site. There, we will need to support:

* multiples GHE accounts
* GHE and regular GH connections living together
* accessing to a private network of GHE instace

Anyway, I was checking if we could advance in one of those point in this PR but it seems that we won't be able to do this yet.

Currently, this is only useful for private installations of RTD source code.

File: docs/custom_installs/github_enterprise.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (gaborbernat) [reply docs/custom_installs/github_enterprise.rst]: okay I can do; my approach was that people in an enterprise env usually just want one VCS, hence why removing gitlab is not needed anymore

File: readthedocs/projects/constants.py (hunk 7 above)
Comment (gaborbernat) [reply readthedocs/projects/constants.py]: no this change is needed, because now we'll format the string to get the regex, hence the double quotation actually keeps it as it was beforehand (aka literal ``{``)

**Comment by gaborbernat** [comment]
@humitos tox organization has a ``fix-lint`` target to setup and run automatically the ``pre-commit`` maybe we should have it here too, but that's part of a different PR/issue

Indeed this PR is useful for private installations of RTD. We can look into supporting more part of further subsequent PRs. I don't have access to multiple instances of these, hence why was not focusing on taking a look at it. But most likely it's something that can be done. 

Additionally made the required change (this way we keep previous settings - ``gitlab`` included, only upgrade the github URL):

```python
from .dev import CommunityDevSettings
_SETTINGS = CommunityDevSettings()

SOCIALACCOUNT_PROVIDERS = _SETTINGS.SOCIALACCOUNT_PROVIDERS
SOCIALACCOUNT_PROVIDERS['github']['GITHUB_URL'] == '{{github_url}}'
```

File: readthedocs/oauth/services/github.py [hunk 8]
```diff
@@ -28,7 +28,7 @@ class GitHubService(Service):
 
     adapter = GitHubOAuth2Adapter
     # TODO replace this with a less naive check
-    url_pattern = re.compile(r'github\.com')
+    url_pattern = re.compile(re.escape(GitHubOAuth2Adapter.web_url))
```
Comment (humitos) [inline readthedocs/oauth/services/github.py]: I just realized that we can simply use `self.adapter` in all this class.

(it's the common usage for all the services implemented in RTD)

**Comment by gaborbernat** [comment]
@humitos done

**Comment by humitos** [comment]
@gaborbernat nice! Thanks. There is a linting error, could you check it?

To me, this PR is ready for merge. Let's wait for another core team member to review it -- @agjohnson was interested on these changes also.

**Comment by RichardLitt** [comment]
You're heard, @gaborbernat. Four days (and over the weekend) isn't a lot of time for us - we'll get to this when we can.

**Comment by agjohnson** [comment]
I'm happy that someone is able to take this on. Thanks for the contribution @gaborbernat !

I'll have more feedback when I can focus on this more, but for starters: this is touching a lot of important code and has no tests. So, fixing that is a good place to continue this work.

I'm going to target a future release, as this isn't quite on our roadmap. I think this is mostly close though. Apologies for any delays responding, off-roadmap PRs have lower priority for us.

**Comment by gaborbernat** [comment]
Just an update. I definitely plan to continue to work on this. Just other things have come up. The fact that the actual code has not tests at the moment means I first need to figure out how this part is intended to be tested. Which makes picking up this non-trivial.

**Comment by gaborbernat** [comment]
Due to lack of resource for working on this, I'll need to retract this PR. Thanks for the collaboration guys.

**Comment by RichardLitt** [comment]
@gaborbernat Thanks for letting us know, and for your work, even if it wasn't merged in the end. :)
